mono/metadata/boehm-gc.c

```diff
@@ -1649,69 +1680,66 @@ handle_data_first_unset (HandleData *handles)
-	guint32 new_size = handles->size * 2; /* always double: we memset to 0 based on this below */
+	return (MonoGCHandle)((uintptr_t)handle & ~(uintptr_t)1);
+}
 
-	/* resize and copy the bitmap */
-	new_bitmap = (guint32 *)g_malloc0 (new_size / CHAR_BIT);
-	memcpy (new_bitmap, handles->bitmap, handles->size / CHAR_BIT);
-	g_free (handles->bitmap);
-	handles->bitmap = new_bitmap;
+static HandleData*
+get_handle_data_from_handle (MonoGCHandle handle)
+{
+	HandleData* handles = (HandleData*)ALIGN_DOWN_TO(handle, HANDLE_DATA_ALIGNMENT);
+	return handles;
+}
 
-	/* resize and copy the entries */
-	if (MONO_GC_HANDLE_TYPE_IS_WEAK (handles->type)) {
-		gpointer *entries;
-		guint16 *domain_ids;
-		gint i;
-		domain_ids = (guint16 *)g_malloc0 (sizeof (*handles->domain_ids) * new_size);
-		entries = (void **)g_malloc0 (sizeof (*handles->entries) * new_size);
-		memcpy (domain_ids, handles->domain_ids, sizeof (*handles->domain_ids) * handles->size);
-		for (i = 0; i < handles->size; ++i) {
-			MonoObject *obj = mono_gc_weak_link_get (&(handles->entries [i]));
-			if (obj) {
-				mono_gc_weak_link_add (&(entries [i]), obj, track);
-				mono_gc_weak_link_remove (&(handles->entries [i]), track);
-			} else {
-				g_assert (!handles->entries [i]);
-			}
-		}
-		g_free (handles->entries);
-		g_free (handles->domain_ids);
-		handles->entries = entries;
-		handles->domain_ids = domain_ids;
-	} else {
-		gpointer *entries;
-		entries = (void **)mono_gc_alloc_fixed (sizeof (*handles->entries) * new_size, NULL, MONO_ROOT_SOURCE_GC_HANDLE, NULL, "GC Handle Table (Boehm)");
-		mono_gc_memmove_aligned (entries, handles->entries, sizeof (*handles->entries) * handles->size);
-		mono_gc_dirty_range (entries, new_size * sizeof (*handles->entries));
-		mono_gc_free_fixed (handles->entries);
-		handles->entries = entries;
-	}
-	handles->slot_hint = handles->size / BITMAP_SIZE;
-	handles->size = new_size;
+static HandleData*
+mono_gchandle_lookup (MonoGCHandle handle, guint* slot)
+{
+	HandleData* handles = get_handle_data_from_handle (handle);
+	if (slot)
+		*slot = (int)(ptrdiff_t)((gpointer*)mono_gchandle_untag_strong (handle) - &handles->entries[0]);
+	return handles;
 }
 
-static guint32
-alloc_handle (HandleData *handles, MonoObject *obj, gboolean track)
+static MonoGCHandle
+alloc_handle (int type, MonoObject *obj, gboolean track)
 {
-	gint slot, i;
-	guint32 res;
+	gint slot, i = -1;
+	MonoGCHandle res;
+	HandleData* handles = gc_handles[type];
 	lock_handles (handles);
-	if (!handles->size)
-		handle_data_alloc_entries (handles);
-	i = handle_data_next_unset (handles);
-	if (i == -1 && handles->slot_hint != 0)
-		i = handle_data_first_unset (handles);
+	if (!handles) {
+		handles = gc_handles[type] = gc_handles_free[type] = handle_data_alloc_entries(type);
+	}
+	handles = gc_handles_free[type];
+	if (handles) {
+		i = handle_data_next_unset(handles);
+		if (i == -1 && handles->slot_hint != 0)
+			i = handle_data_first_unset(handles);
+	}
 	if (i == -1) {
```

Comment:
I feel like this is just an else condition to `if(gc_handles_free[type])`

if `gc_handles_free[type]` is not null and you did not find a slot, something went wrong?

Reply:
Also before we allocate a new set of entries, we need to check all of the handles->next_free links?

Reply:
> if gc_handles_free[type] is not null and you did not find a slot, something went wrong?

Yes, I think you are right. The contract is, a chunk in the freelist by definition has free slots. I could change this to an assert.

Reply:
> I feel like this is just an else condition to if(gc_handles_free[type])
if gc_handles_free[type] is not null and you did not find a slot, something went wrong?
Also before we allocate a new set of entries, we need to check all of the handles->next_free links?

I think the first part of your questions is correct. Any chunk in the freelist will have a free slot. I am changing the code to get a slot or assert/fail as something is wrong. Given this, we don't ever need to iterate the free list. And, I changed the code so we only every allocate a new chunk if `handles == NULL`, i.e. the free list is empty.